### PR TITLE
Update ClusteringMethod.py

### LINKED ISSUE #2575: ClusteringMethods don't need to return details
**Expected behavior**

Everything returned by `ClusteringMethod.__call__` contains information that is used.

**Actual behavior**

`ClusteringMethod.__call__` returns ``clusters, details``. ``details`` is always an empty dictionary. It's not used by ``cluster`` and is not even in the documentation:

https://github.com/MDAnalysis/mdanalysis/blob/9bcf6f4c118e1ea137e8514bd60cbd1cd1972062/package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py#L77-L91

The documentation is also inaccurate: ``clusters`` is an array of the centroid frame indices of length \#elements, as returned by ``encode_centroid_info``:

https://github.com/MDAnalysis/mdanalysis/blob/9bcf6f4c118e1ea137e8514bd60cbd1cd1972062/package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py#L57-L66

**Code to reproduce the behavior**

As shown below, ``cluster`` does not use the empty ``details`` dictionary.

https://github.com/MDAnalysis/mdanalysis/blob/9bcf6f4c118e1ea137e8514bd60cbd1cd1972062/package/MDAnalysis/analysis/encore/clustering/cluster.py#L234-L235

**Currently version of MDAnalysis**

- Which version are you using? (run `python -c "import MDAnalysis as mda; print(mda.__version__)"`) 0.20.2-dev
- Which version of Python (`python -V`)?
- Which operating system?

**Suggested fix**
Remove ``details`` from all the ``ClusteringMethod``s and update the documentation. Run the tests to make sure it doesn't affect anything else.

### PR DESCRIPTION
**Pull request by shfrz**
Fixes #2575

Changes made in this Pull Request:
 -corrected documentation
-removed `details` from all the `ClusteringMethod`s
-updated `test_encore.py`
-updated `ClusterCollection.py`

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [ ] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py [hunk 1]
```diff
@@ -156,8 +156,7 @@ def __call__(self, distance_matrix):
             max_iterations = self.max_iter,
             convergence = self.convergence_iter,
             noise=int(self.add_noise))
-        details = {}
-        return clusters, details
+       return clusters
```
Comment (IAlibay) [inline package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py]: Looks like the indentation here is off by a space.

Comment (shfrz) [reply package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py]: Thank you @IAlibay  for pointing that out!

Comment (shfrz) [reply package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py]: > Just providing a cursory review for now as it looks like the tests are failing. There are multiple places in the tests that expect `details` to be returned, you will probably need to account for these.
> 
> Example:
> 
> https://github.com/MDAnalysis/mdanalysis/blob/1f4e492662955e483fc98982728404149f7a1deb/testsuite/MDAnalysisTests/analysis/test_encore.py#L264-L271

I changed the test, could you please review? Thanks!

Comment (IAlibay) [reply package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py]: @shfrz I'm not seeing any changes to `test_encore.py`. Did you push the changes?

Comment (shfrz) [reply package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py]: https://github.com/MDAnalysis/mdanalysis/pull/2588

plz see

Comment (IAlibay) [reply package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py]: @shfrz These changes need to all be within the same pull request. This is particularly important for tests as we need to make sure that Travis properly runs the tests before merging.

Comment (shfrz) [reply package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py]: @IAlibay Did as directed.

**Comment by orbeckst** [comment]
@shfrz you can also run the tests locally, which is much faster than waiting for the CI. There's a howto in https://www.mdanalysis.org/UserGuide/testing.html.

We generally don't look at PRs in detail until the tests run. If you have questions on your work before tests, ask in the PR comments.

File: package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py (hunk 1 above)
Comment (IAlibay) [reply package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py]: @shfrz The are ~ 18 tests still falling here. As initially mentioned above, there are many occurrences of details being expected in the tests, not just the one case I highlighted as an example.

As mentioned by @orbeckst, I would suggest running these locally and seeing if removing the details calls fixes the tests.

Let us know if you need any help with running tests locally / diagnosing pytest failures.

**Comment by lilyminium** [comment]
@shfrz If you [run the tests locally](https://www.mdanalysis.org/UserGuide/testing.html#running-the-test-suite), you can re-run `analysis/test_encore.py` so you can quickly find out which tests are failing. In this case you cannot construct a `ClusterCollection` properly. You'll have to work out why, and then likely change something in `ClusterCollection.__init__` to fix it.

**Comment by lilyminium** [comment]
@shfrz To make developing for MDAnalysis easier, I would [create a local development environment](https://www.mdanalysis.org/UserGuide/contributing_code.html#creating-a-development-environment), [download and compile MDAnalysis](https://www.mdanalysis.org/UserGuide/contributing_code.html#building-mdanalysis), [update the code on my *local* machine](https://www.mdanalysis.org/UserGuide/contributing_code.html#writing-new-code), and [run tests locally](https://www.mdanalysis.org/UserGuide/testing.html#running-the-test-suite) *before* pushing to this pull request. This might involve making many small changes to the code and running the tests (``pytest testsuite/MDAnalysisTests/analysis/test_encore.py``) over and over again, until they pass. I would only commit and push my changes once tests were passing. This saves both your time (testing only that one file might take 5 minutes, compared to hours for the whole test suite on Travis) and other developers' time, as you re-join the queue for Travis testing with each most recent push.

Please don't hesitate to ping me or another developer if you have any questions, especially on how to set up a development environment with local tests, or how the MDAnalysis code works.

**Comment by orbeckst** [comment]
@shfrz do you edit the files locally or through the GitHub edit functionality? We _strongly_ recommend the former together with the local testing that @lilyminium explained in https://github.com/MDAnalysis/mdanalysis/pull/2585#issuecomment-596807694 . One reason why we make you work on a PR is so that you learn how to work with our code and build system and so that we can judge if you are able to work productively.

**Comment by shfrz** [comment]
@lilyminium I tried running the tests locally but it ain't showing me the same errors as travis.

**Comment by lilyminium** [comment]
Perhaps you are testing against the stable version of the code, @shfrz. Do you use ``virtualenv`` or ``conda`` to manage your environments? Have you activated those before you run tests? You may also need to check out the right branch.

**Comment by shfrz** [comment]
> Perhaps you are testing against the stable version of the code, @shfrz. Do you use `virtualenv` or `conda` to manage your environments? Have you activated those before you run tests? You may also need to check out the right branch.


Thanks! I solved this issue

**Comment by shfrz** [comment]
> @shfrz do you edit the files locally or through the GitHub edit functionality? We _strongly_ recommend the former together with the local testing that @lilyminium explained in [#2585 (comment)](https://github.com/MDAnalysis/mdanalysis/pull/2585#issuecomment-596807694) . One reason why we make you work on a PR is so that you learn how to work with our code and build system and so that we can judge if you are able to work productively.

I have configured the local environment successfully. Also, indeed the tests are super quick to run locally.
Thanks a lot!

**Comment by shfrz** [comment]
@lilyminium, in this particular issue while running pytest ~/test_encore.py should I try to make sure that only ClusterCollection.py has no errors?

**Comment by lilyminium** [comment]
Great to hear, @shfrz! 

> in this particular issue while running pytest ~/test_encore.py should I try to make sure that only ClusterCollection.py has no errors?

All the tests need to pass for your pull request to be merged into the regular code-base.

**Comment by shfrz** [comment]
@lilyminium 
` self = <MDAnalysisTests.analysis.test_encore.TestEncore object at 0x00000211BC75D610>, ens1 = <Universe with 3341 atoms>

    def test_rmsd_matrix_with_superimposition_custom_weights(self, ens1):
>       conf_dist_matrix = encore.confdistmatrix.conformational_distance_matrix(
            ens1,
            encore.confdistmatrix.set_rmsd_matrix_elements,
            select="name CA",
            pairwise_align=True,
            weights='mass',
            n_jobs=1)
E       TypeError: conformational_distance_matrix() got an unexpected keyword argument 'select'
`

I'm getting this error again and again, I tried to fix this so many times and also checked `confdistmatrix.py`, but it seemed fine
any clue?

File: package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py (hunk 1 above)
Comment (shfrz) [reply package/MDAnalysis/analysis/encore/clustering/ClusteringMethod.py]: @IAlibay @lilyminium @orbeckst  , all tests are passing now except for those which are independent of this particular issue!
Shouldn't the changes be merged now and the issue be closed as the tests directly related to this issue are all passing?

Thanks!

**Comment by lilyminium** [comment]
@shfrz It sounds like you have an old version of the code -- [try rebasing from develop](https://www.mdanalysis.org/UserGuide/contributing_code.html#rebasing-your-code).

>  all tests are passing now except for those which are independent of this particular issue!

While tests may be passing for you locally, you haven't pushed your changes to this pull request, so we can't see that. Again, all tests need to pass so we can make sure the code is stable.

**Comment by RMeli** [comment]
> in this particular issue while running pytest ~/test_encore.py should I try to make sure that only ClusterCollection.py has no errors?

As @lilyminium already pointed out, all tests need to pass for your PR to be merged. However, to speedup development, you can [run specific tests](https://www.mdanalysis.org/UserGuide/testing.html#running-the-test-suite) for the module you are working on and then run all the tests just before opening the PR (to make sure your changes didn't broke something else, which will need fixing).

**Comment by shfrz** [comment]
@lilyminium , no that isn't the case

Capture_test


I am trying to solve this issue continuously since I ran the tests locally, I wonder why `select` is an unexpected keyword argument as it is already assigned in ` confdistmatrix.conformational_distance_matrix`
?

Removing or adding `details` has no effect on this issue.

**Comment by lilyminium** [comment]
@shfrz This is from the changes in #2528. I think it's likely you are not testing against the development version of the code, because the tests are clearly up-to-date. If you run (in Python)

```python
import MDAnalysis as mda
mda.__path__
```

Does it give you the path to your local repo?

**Comment by shfrz** [comment]
@lilyminium , I am indeed testing against the developmental version!
`pytest mdanalysis\testsuite\MDAnalysisTests\analysis\test_encore.py` this is exactly the cloned develop repo....

**Comment by shfrz** [comment]
@RMeli , the tests for the specific module that I am working on are passing successfully, but when I run test for `test_encore.py`, I get some errors that aren't related to the module..

**Comment by shfrz** [comment]
@lilyminium , Even the Travis Cl is throwing the same error as my local setup, 
I think changing `select` to `selection` works fine!

**Comment by shfrz** [comment]
Hi @orbeckst , the tests passed locally 💯 
`pytest ~\test_encore.py`

Capturetest2


Hence, I'm committing...

Thanks!

File: testsuite/MDAnalysisTests/analysis/test_encore.py [hunk 2]
```diff
@@ -155,15 +155,15 @@ def test_rmsd_matrix_with_superimposition_custom_weights(self, ens1):
         conf_dist_matrix = encore.confdistmatrix.conformational_distance_matrix(
             ens1,
             encore.confdistmatrix.set_rmsd_matrix_elements,
-            select="name CA",
+            selection="name CA",
```
Comment (RMeli) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: As already pointed out by @lilyminium, `selection` has been changed to `select` in #2528

File: testsuite/MDAnalysisTests/analysis/test_encore.py [hunk 3]
```diff
@@ -155,15 +155,15 @@ def test_rmsd_matrix_with_superimposition_custom_weights(self, ens1):
...
             pairwise_align=True,
             weights='mass',
             n_jobs=1)
 
         conf_dist_matrix_custom = encore.confdistmatrix.conformational_distance_matrix(
             ens1,
             encore.confdistmatrix.set_rmsd_matrix_elements,
-            select="name CA",
+            selection="name CA",
```
Comment (RMeli) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: `selection` has been changed to `select` in #2528

File: testsuite/MDAnalysisTests/analysis/test_encore.py [hunk 4]
```diff
@@ -182,7 +182,7 @@ def test_rmsd_matrix_without_superimposition(self, ens1):
         confdist_matrix = encore.confdistmatrix.conformational_distance_matrix(
             ens1,
             encore.confdistmatrix.set_rmsd_matrix_elements,
-            select=selection_string,
+            selection=selection_string,
```
Comment (RMeli) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: `selection` has been changed to `select` in #2528

File: testsuite/MDAnalysisTests/analysis/test_encore.py [hunk 5]
```diff
@@ -285,7 +285,7 @@ def test_dres_to_self(self, ens1):
                             err_msg="Dim. Reduction Ensemble Similarity to itself not zero: {0:f}".format(result_value))
 
     def test_dres(self, ens1, ens2):
-        results, details = encore.dres([ens1, ens2], select="name CA and resnum 1-10")
+        results, details = encore.dres([ens1, ens2], selection="name CA and resnum 1-10")
```
Comment (RMeli) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: `selection` has been changed to `select` in #2528

File: testsuite/MDAnalysisTests/analysis/test_encore.py [hunk 6]
```diff
@@ -353,7 +353,7 @@ def test_ces_error_estimation(self, ens1):
                                       estimate_error = True,
                                       bootstrapping_samples=10,
                                       clustering_method=encore.AffinityPropagationNative(preference=-2.0),
-                                      select="name CA and resnum 1-10")
+                                      selection="name CA and resnum 1-10")
```
Comment (RMeli) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: `selection` has been changed to `select` in #2528

File: testsuite/MDAnalysisTests/analysis/test_encore.py [hunk 7]
```diff
@@ -375,7 +375,7 @@ def test_ces_error_estimation_ensemble_bootstrap(self, ens1):
                                       estimate_error = True,
                                       bootstrapping_samples=10,
                                       clustering_method=encore.KMeans(n_clusters=2),
-                                      select="name CA and resnum 1-10")
+                                      selection="name CA and resnum 1-10")
```
Comment (RMeli) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: `selection` has been changed to `select` in #2528

File: testsuite/MDAnalysisTests/analysis/test_encore.py [hunk 8]
```diff
@@ -389,7 +389,7 @@ def test_dres_error_estimation(self, ens1):
         stdev_upper_bound = 0.2
         averages, stdevs = encore.dres([ens1, ens1], estimate_error = True,
                                        bootstrapping_samples=10,
-                                       select="name CA and resnum 1-10")
+                                       selection="name CA and resnum 1-10")
```
Comment (RMeli) [inline testsuite/MDAnalysisTests/analysis/test_encore.py]: `selection` has been changed to `select` in #2528

**Comment by shfrz** [comment]
Hi @RMeli , @lilyminium 
I am using the dev version for both MDAnalysis and  MDAnalysisTests

Capture5

Capture6

When I run the tests locally I get no errors 

Capturetest2

But Travis Cl is showing errors here..
Any clue why?

**Comment by lilyminium** [comment]
@shfrz that is puzzling! Is the code in your local repository the same as the code in this pull request? When you type `git log` into terminal, could you show us as much of the output as possible?

**Comment by shfrz** [comment]
> @shfrz that is puzzling! Is the code in your local repository the same as the code in this pull request? When you type `git log` into terminal, could you show us as much of the output as possible?

@lilyminium 
gitlog1

**Comment by RMeli** [comment]
Strange... How are you running the tests? Any chance you are using an IDE which points to a different Python environment?

**Comment by shfrz** [comment]
> Strange... How are you running the tests? Any chance you are using an IDE which points to a different Python environment?

@RMeli , no I am using the Anaconda Prompt cmd

**Comment by shfrz** [comment]
@RMeli , any clue how to fix this discrepancy between local test results and Travis Cl?

**Comment by orbeckst** [comment]
When I look at the Travis tests (eg https://travis-ci.com/github/MDAnalysis/mdanalysis/jobs/297187652) I see errors that look related to your changes, e.g.,
```
>               self.clusters.append(

                    Cluster(elem_list=this_array[0], idn=idn, centroid=c,
                            metadata=this_metadata))
E                           AttributeError: 'ClusterCollection' object has no attribute 'clusters'

package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py:205: AttributeError

...

>       if not len(set((type(el) for el in elements_array))) == 1:
E       TypeError: iteration over a 0-d array

package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py:192: TypeError
...
```

**Comment by shfrz** [comment]
> When I look at the Travis tests (eg https://travis-ci.com/github/MDAnalysis/mdanalysis/jobs/297187652) I see errors that look related to your changes, e.g.,
> 
> ```
> >               self.clusters.append(
> 
>                     Cluster(elem_list=this_array[0], idn=idn, centroid=c,
>                             metadata=this_metadata))
> E                           AttributeError: 'ClusterCollection' object has no attribute 'clusters'
> 
> package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py:205: AttributeError
> 
> ...
> 
> >       if not len(set((type(el) for el in elements_array))) == 1:
> E       TypeError: iteration over a 0-d array
> 
> package/MDAnalysis/analysis/encore/clustering/ClusterCollection.py:192: TypeError
> ...
> ```

Yes, I did some research and found out that conda behaves differently with regard to packages installed in virtual environments and root. I am fixing that. 
I think while playing with mdanalysis during initial phase, I installed some unneeded stuff!

**Comment by shfrz** [comment]
@orbeckst , Thanks a ton!

**Comment by shfrz** [comment]
This PR is stacking commits unusually, I am closing this PR and opening a new one soon!

**Comment by zemanj** [comment]
> This PR is stacking commits unusually

What do you mean by this? We're using git here. Please don't open a new PR if this one can be fixed. That way, we can keep the discussion in one place.

**Comment by zemanj** [comment]
BTW are you sure you built and installed your changes locally before running the tests?

**Comment by orbeckst** [comment]
PLEASE don't just close the PR. With `git` there is _always_ a way to fix it.

Tell us what you want to accomplish. We can probably tell you what to do.

**Comment by shfrz** [comment]
@orbeckst @zemanj ,
I think my conda environment got mixed up! I was getting different test results

Here is what I did :
1. Cloned repository using git
2. Opened the patch-2 using Github Desktop
3. made changes in files using sublime Text Editor
4. ran tests using pytest

Is this the right approach?
Any suggestions are most welcome!

Thanks.

Regards,
Faraaz

**Comment by zemanj** [comment]
I don't think that saving the files is enough as these changes will live in your local repo but won't be reflected in the installed version of MDAnalysis. I am quite sureou will have to build & install the package locally

**Comment by shfrz** [comment]
Hi @zemanj, 
I already have the developmental version installed for both MDAnalysis as well as the MDAnalysisTestsuite 

@orbeckst 

I have worked it out now pulling a few all-nighters, some dry eyes....

**Comment by orbeckst** [comment]
I have never used GitHub desktop so I cannot quite comment on what you're doing; I only ever use `git` from the commandline. I would follow https://github.com/MDAnalysis/mdanalysis/pull/2585#issuecomment-596807694 and work with the **conda environment**.

1. [Fork](https://guides.github.com/activities/forking/) MDAnalysis: you should now have a repository https://github.com/shfrz/mdanalysis
2. Clone your repo with
     ```
     git clone https://github.com/shfrz/mdanalysis.git
     cd mdanalysis
     git remote add upstream https://github.com/MDAnalysis/mdanalysis
     ```
3. [Create the conda environment](https://www.mdanalysis.org/UserGuide/contributing_code.html#with-conda)
4. Make sure you can [build MDAnalysis](https://www.mdanalysis.org/UserGuide/contributing_code.html#building-mdanalysis)
5. [Create a branch](https://www.mdanalysis.org/UserGuide/contributing_code.html#creating-a-branch) for your changes, e.g.
    ```
    git checkout -b update-clustering-method-2575
    ```
6. Work on your code, `git add` your changes, [`git commit`](https://www.mdanalysis.org/UserGuide/contributing_code.html#committing-your-code); also run the tests locally.
7. [Push](https://www.mdanalysis.org/UserGuide/contributing_code.html#pushing-your-code-to-github) your changes to your fork:
    ```
    git push origin update-clustering-method-2575
    ```
8. [Create a pull request](https://www.mdanalysis.org/UserGuide/contributing_code.html#creating-a-pull-request)

**Comment by shfrz** [comment]
@orbeckst , Now I am getting the same errors as Travis would generate!
I tried removing `details` from ClusteringMethods.py but it generates errors in ClusterCollection.py,
After fixing errors there, it seems that fixing one error generates another error and so on.
So what should we do?

Thanks again!

**Comment by lilyminium** [comment]
@shfrz In [the issue](https://github.com/MDAnalysis/mdanalysis/issues/2575#issue-573645989) I gave an example of ``cluster`` not using the empty ``details`` dictionary. Have you looked at that code and how it might change if ClusteringMethods do not return ``details``?

**Comment by shfrz** [comment]
@lilyminium , how can I commit to this PR from a new branch?

**Comment by lilyminium** [comment]
@shfrz You appear to have deleted the fork that this PR came from, so you can't. You'll have to open a new one. If you want to keep the work you have done here, you'll need to pull the PR (command-line instructions given, since I have also never used Github Desktop):
```
git fetch upstream refs/pull/2585/head:my-new-branch-name
```
check out the new branch
```
git checkout my-new-branch-name
```
and push to remote again.
```
git push origin my-new-branch-name
```

Why did you delete your fork of the repository?

**Comment by shfrz** [comment]
> @shfrz In [the issue](https://github.com/MDAnalysis/mdanalysis/issues/2575#issue-573645989) I gave an example of `cluster` not using the empty `details` dictionary. Have you looked at that code and how it might change if ClusteringMethods do not return `details`?

@lilyminium, I removed the details from ClusteringMethods, but the tests are failing again, I also looked at the code but it doesn't fix that...
What to do now?
Can we discuss this on one-on-one conversation to fix this issue faster?
Thanks!

**Pull request closed without merging**